src/main/java/gregtech/common/covers/detector/CoverDetectorEnergyAdvanced.java

```diff
@@ -66,166 +61,222 @@ public EnumActionResult onScrewdriverClick(EntityPlayer playerIn, EnumHand hand,
-                .setTooltipHoverString("cover.advanced_energy_detector.invert_tooltip")
-        );
+        // get/set min EU
+        group.addWidget(new LabelWidget(10, 5 + (SIZE + PADDING), "cover.advanced_energy_detector.min"));
+        group.addWidget(new ImageWidget(72, (SIZE + PADDING), 8 * SIZE, SIZE, GuiTextures.DISPLAY));
 
         // get/set max EU
         group.addWidget(new LabelWidget(10, 5 + 2 * (SIZE + PADDING), "cover.advanced_energy_detector.max"));
-        group.addWidget(new ImageWidget(68, 2 * (SIZE + PADDING), 4 * SIZE, SIZE, GuiTextures.DISPLAY));
-        group.addWidget(new TextFieldWidget2(72, 5 + 2 * (SIZE + PADDING), 4 * SIZE, SIZE,
-                this::getMaxValue, this::setMaxValue)
-                    .setMaxLength(19)
-                    .setAllowedChars(ALLOWED_CHARS)
-                    .setPostFix(" EU")
+        group.addWidget(new ImageWidget(72, 2 * (SIZE + PADDING), 8 * SIZE, SIZE, GuiTextures.DISPLAY));
+
+        updateSyncedWidgets(); // update widgets on UI creation
+
+        // change modes between percent and discrete EU
+        group.addWidget(new LabelWidget(10, 5 + 3 * (SIZE + PADDING), "cover.advanced_energy_detector.modes_label"));
+        group.addWidget(new CycleButtonWidget(72, 3 * (SIZE + PADDING), 4 * SIZE, SIZE, this::isUsePercent, this::setUsePercent,
+                "cover.advanced_energy_detector.mode_eu", "cover.advanced_energy_detector.mode_percent")
+                .setTooltipHoverString("cover.advanced_energy_detector.modes_tooltip")
         );
 
-        // get/set min EU
-        group.addWidget(new LabelWidget(10, 5 + 3 * (SIZE + PADDING), "cover.advanced_energy_detector.min"));
-        group.addWidget(new ImageWidget(68, 3 * (SIZE + PADDING), 4 * SIZE, SIZE, GuiTextures.DISPLAY));
-        group.addWidget(new TextFieldWidget2(72, 5 + 3 * (SIZE + PADDING), 4 * SIZE, SIZE,
-                this::getMinValue, this::setMinValue)
-                    .setMaxLength(19)
-                    .setAllowedChars(ALLOWED_CHARS)
-                    .setPostFix(" EU")
+        // invert logic button
+        group.addWidget(new LabelWidget(10, 5 + 4 * (SIZE + PADDING), "cover.advanced_energy_detector.invert_label"));
+        group.addWidget(new CycleButtonWidget(72, 4 * (SIZE + PADDING), 4 * SIZE, SIZE, this::isInverted, this::setInverted,
+                "cover.advanced_energy_detector.normal", "cover.advanced_energy_detector.inverted")
+                .setTooltipHoverString("cover.advanced_energy_detector.invert_tooltip")
         );
 
-        return ModularUI.builder(GuiTextures.BACKGROUND, 176, 108)
+        return ModularUI.builder(GuiTextures.BACKGROUND, 176 + (3 * SIZE), 108 + (SIZE))
                 .widget(group)
+                .widget(widgetsToUpdate) // add synced widgets
                 .build(this, player);
     }
 
+    private WidgetGroup constructWidgetsToUpdate() {
```

Comment:
Given that GUI is always constructed anew when opened. It does not really makes sense to construct this part beforehand and then with difficulties (and bugs) partially update it when opening GUI.

Sure you need to update it on behalf of changes in GUI but construction should be done at GUI creation too. Also you can save just the two Widgets you really need and save yourself checking for `instanceof`

Reply:
the call has been moved back into `createUI()`, and the `instanceof` check has been removed. (https://github.com/GregTechCEu/GregTech/pull/1511/commits/dd65ee3cfb8f99b47c57a0cf8d1da8d58b66b562)